aggregator/entry.go

```diff
@@ -0,0 +1,243 @@
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package aggregator
+
+import (
+	"container/list"
+	"errors"
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/m3db/m3metrics/metric"
+	"github.com/m3db/m3metrics/metric/unaggregated"
+	"github.com/m3db/m3metrics/policy"
+)
+
+var (
+	errEntryClosed = errors.New("entry is closed")
+)
+
+// Entry stores metadata about current policies and aggregations
+// for a metric. Note since the metric namespace is part of the
+// metric ID, we only need to keep track of the policy version
+// to detect policy changes
+type Entry struct {
+	sync.RWMutex
+
+	opts           Options                         // options
+	lists          *MetricLists                    // metric lists
+	version        int                             // current policy version
+	numWriters     int32                           // number of writers writing to this entry
+	lastAccessInNs int64                           // last access time
+	aggregations   map[policy.Policy]*list.Element // aggregations at different resolutions
+	closed         bool                            // whether the entry is closed
+}
+
+// NewEntry creates a new entry
+func NewEntry(lists *MetricLists, opts Options) *Entry {
+	e := &Entry{
+		opts:         opts,
+		aggregations: make(map[policy.Policy]*list.Element),
+	}
+	e.ResetSetData(lists)
+	return e
+}
+
+// IncWriter increases the writer count
+func (e *Entry) IncWriter() { atomic.AddInt32(&e.numWriters, 1) }
+
+// DecWriter decreases the writer count
+func (e *Entry) DecWriter() { atomic.AddInt32(&e.numWriters, -1) }
+
+// ResetSetData resets the entry and sets initial data
+func (e *Entry) ResetSetData(lists *MetricLists) {
+	e.closed = false
+	e.lists = lists
+	e.version = policy.InitPolicyVersion
+	e.numWriters = 0
+	e.recordLastAccessed(e.opts.ClockOptions().NowFn()())
+}
+
+// AddMetricWithPolicies adds a metric along with applicable policies
+func (e *Entry) AddMetricWithPolicies(
```

Comment:
JFYI: I found it somewhat more elegant to write this kind of code (RLock -> Check -> RUnlock -> Lock -> Check -> Action -> Unlock) by using `sync.Locker` objects, specifically by factoring out the common part into a function that accepts a `sync.Locker` to avoid repetition.